### PR TITLE
Code cleanup and documentation work. Unfinished.

### PR DESCRIPTION
**Pull request by d0nut-grapl**
### Which issue does this PR correspond to?

#255 

### What changes does this PR make to Grapl? Why?

This PR ( _will_ ) contain three major types of changes:
1. Rust-Doc documentation that fills out the docs generated by invoking `cargo doc`
2. (_Not included, yet_) Inline code comments designed to help clarify complicated code regions
3. Code changes and optimizations as I see them that are designed to **_not_** change the behavior of Grapl except to reduce clones, reduce code-count, and use utilities already available in the code-base (e.g. `grapl_config::is_local` instead of checking the `env::var` yourself).

### How were these changes tested?

Code was build and dobi unit tests were run.

### CODE REVIEW

File: src/rust/graph-generator-lib/src/lib.rs [hunk 1]
```diff
@@ -30,26 +30,6 @@ use sqs_lambda::event_handler::EventHandler;
...
-        self.buffer.clear();
-
-        let mut body = Cursor::new(&body);
-
-        zstd::stream::copy_decode(&mut body, &mut self.buffer)?;
-
-        Ok(serde_json::from_slice(&self.buffer)?)
-    }
-}
-
```
Comment (d0nut-grapl) [inline src/rust/graph-generator-lib/src/lib.rs:52]: This was unused and seems to have a sister implementation in [generic-subgraph-generator](https://github.com/grapl-security/grapl/blob/staging/src/rust/generic-subgraph-generator/src/main.rs#L628) where it is actually used.

File: src/rust/graph-generator-lib/src/lib.rs [hunk 2]
```diff
@@ -290,11 +270,7 @@ fn handler<
     let (tx, rx) = std::sync::mpsc::sync_channel(10);
     let completed_tx = tx.clone();
 
-    let generator = generator.clone();
```
Comment (d0nut-grapl) [inline src/rust/graph-generator-lib/src/lib.rs:293]: This all compiles ; just unnecessary `.clone()` invocations

File: src/rust/sysmon-subgraph-generator/src/main.rs [hunk 3]
```diff
@@ -505,18 +547,6 @@ where
...
-    C: Cache + Clone + Send + Sync + 'static,
-{
-    fn clone(&self) -> Self {
-        Self {
-            cache: self.cache.clone(),
-            metrics: self.metrics.clone(),
-        }
-    }
-}
-
```
Comment (d0nut-grapl) [inline src/rust/sysmon-subgraph-generator/src/main.rs:519]: Not sure why `Clone` was explicitly implemented but `#[derive(Clone)]` worked fine. As long as all the members implement `Clone`, the parent should be able to implement `Clone` without explicit intervention.

File: src/rust/sysmon-subgraph-generator/src/main.rs [hunk 4]
```diff
@@ -673,13 +703,16 @@ async fn main() -> Result<(), Box<dyn std::error::Error>> {
     };
 
     if grapl_config::is_local() {
-        let generator = SysmonSubgraphGenerator::new(NopCache {}, metrics);
```
Comment (d0nut-grapl) [inline src/rust/sysmon-subgraph-generator/src/main.rs:676]: I could separate the `let generator = ...;` line out again if we want, but the main change is just calling `run_graph_generator` which should be best practice here. 

I really tried to find a way to do the following:


```
// ...

let cache = if grapl_config::is_local() {
    NopCache {}
} else {
    event_cache().await
};

run_graph_generator(SysmonSubgraphGenerator::new(cache, metrics), ZstdDecoder::default()).await;
```

But I was unable to figure out how to make Rust happy with that. If anyone knows of a way that doesn't look like complete gibberish or require some crazy type work, I'd love to see it. I hate this branching pattern.

My suspicion is we could use `Box<dyn Cache ...>` but I've not that much experience with `Box` and not sure if there is a way to pass the value of a `Box` into `SysmonSubgraphGenerator` without changing the type on the `:;new` fn.

**Review by jgrillo-grapl — APPROVED** [review]
👍

File: src/rust/generic-subgraph-generator/src/models/network/inbound.rs [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+            .build()?;
+
+        let src_ip = IpAddressBuilder::default()
+            .ip_address(conn_log.src_ip_addr.clone())
+            .last_seen_timestamp(conn_log.timestamp)
+            .build()?;
+
+        let src_port = IpPortBuilder::default()
+            .ip_address(conn_log.src_ip_addr.clone())
+            .protocol(conn_log.protocol.clone())
```
Comment (d0nut-grapl) [inline src/rust/generic-subgraph-generator/src/models/network/inbound.rs:64]: As it turns out, we were missing the `.protocol()` line here as caught by unit testing :D

File: src/rust/generic-subgraph-generator/src/models/process/port_bind.rs [hunk 6]
```diff
@@ -0,0 +1,9 @@
+use serde::{Deserialize, Serialize};
+
+#[derive(Clone, Debug, Hash, Serialize, Deserialize)]
+pub struct ProcessPortBindLog {
```
Comment (d0nut-grapl) [inline src/rust/generic-subgraph-generator/src/models/process/port_bind.rs:4]: This was previously unimplemented. I've decided to leave it as such as it's not exactly in the scope of this diff to add functionality.

File: src/rust/generic-subgraph-generator/src/tests.rs [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+
+    let mut generator = GenericSubgraphGenerator::new(NopCache {});
+
+    let mut event_deserializer = ZstdJsonDecoder::default();
+
+    let generic_events: Vec<GenericEvent> = event_deserializer
+        .decode(raw_test_data)
+        .expect("Failed to deserialize events.");
+
+    let (subgraph, identities, failed) = generator.convert_events_to_subgraph(generic_events).await;
```
Comment (d0nut-grapl) [inline src/rust/generic-subgraph-generator/src/tests.rs:45]: This caught an issue in Inbound Connection subgraph generation :)

File: src/rust/generic-subgraph-generator/test_data/events.json [hunk 8]
```diff
@@ -0,0 +1,73 @@
+[{
```
Comment (d0nut-grapl) [inline src/rust/generic-subgraph-generator/test_data/events.json:1]: We attempt to use every implemented node type in generic-subgraph-generator in this example log. It tells the story of a process (`example_parent.exe`), spawning a child process (`example.exe`), which then reads sensitive data from Chrome, writes a windows batch file (`run_me.bat`), executes the file which then reaches out to an external IP address, and that IP address then calls back to the original child process (`example.exe`).

Just a note regarding `compressed_events.zstd`:

`compressed_events.zstd` is generated from this file by running `zstd events.json -o compressed_events.zstd`.

File: src/rust/graph-generator-lib/src/aws.rs [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+        .collect();
+
+    info!("Initial Events {:?}", initial_events);
+
+    let (tx, rx) = std::sync::mpsc::sync_channel(10);
+
+    let t = std::thread::spawn(move || {
+        // lambda_runtime uses tokio::runtime::Runtime internally so we need to use
+        // tokio_compat::run_std if we want to invoke a new async task
+        tokio_compat::run_std(run_async_generator_handler(
```
Comment (d0nut-grapl) [inline src/rust/graph-generator-lib/src/aws.rs:70]: This used to be a closure but it was very hard to parse so I moved that out into its own function.

File: src/rust/graph-generator-lib/src/aws.rs [hunk 10]
```diff
@@ -0,0 +1,221 @@
...
+    });
+
+    info!("Checking acks");
+
+    for received_message in &rx {
+        info!("Acking event: {}", &received_message);
+
+        initial_events.remove(&received_message);
+
+        if received_message == "Completed" {
```
Comment (d0nut-grapl) [inline src/rust/graph-generator-lib/src/aws.rs:86]: I was going to change this originally but was more focused on other parts of this diff. This should not be a String comparison but an enum should be used to track state here.

File: src/rust/graph-generator-lib/src/aws.rs [hunk 11]
```diff
@@ -0,0 +1,221 @@
+use crate::serialization::SubgraphSerializer;
```
Comment (d0nut-grapl) [inline src/rust/graph-generator-lib/src/aws.rs:1]: This file contains code that is specific to running Grapl in production (on aws).

File: src/rust/graph-generator-lib/src/local.rs [hunk 12]
```diff
@@ -0,0 +1,197 @@
+use crate::serialization::SubgraphSerializer;
```
Comment (d0nut-grapl) [inline src/rust/graph-generator-lib/src/local.rs:1]: This file contains code that is specific to running grapl locally (or, more specifically, "not on aws"). It performs extra tasks like waiting for S3 and SQS to be available (via localstack/minio) and feeding events to the generator.

File: src/rust/generic-subgraph-generator/src/tests.rs [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+            report
+        );
+    }
+}
+
+async fn read_test_data_to_string(filename: &str) -> Result<String> {
+    let data = read_test_data(filename).await?;
+
+    String::from_utf8(data)
+        .map_err(|utf8_error| tokio::io::Error::new(tokio::io::ErrorKind::Other, utf8_error))
```
Comment (inickles-grapl) [inline src/rust/generic-subgraph-generator/src/tests.rs:58]: that's what i'm talkin about 😎

**Review by inickles-grapl — APPROVED** [review]

File: src/rust/generic-subgraph-generator/src/models/process/port_bind.rs (hunk 6 above)
Comment (jgrillo-grapl) [reply src/rust/generic-subgraph-generator/src/models/process/port_bind.rs:4]: Might be worth adding an issue to track this?

File: src/rust/generic-subgraph-generator/test_data/events.json (hunk 8 above)
Comment (jgrillo-grapl) [reply src/rust/generic-subgraph-generator/test_data/events.json:1]: 💯

**Review by jgrillo-grapl — APPROVED** [review]
This looks like a substantial improvement. 👍

**Review by jgrillo-grapl — APPROVED** [review]
👍

File: src/rust/generic-subgraph-generator/src/models/file/write.rs [hunk 14]
```diff
@@ -0,0 +1,48 @@
...
+    path: String,
+    hostname: String,
+    timestamp: u64,
+}
+
+impl TryFrom<FileWrite> for Graph {
+    type Error = String;
+
+    fn try_from(file_write: FileWrite) -> Result<Self, Self::Error> {
+        let deleter = ProcessBuilder::default()
```
Comment (colin-grapl) [inline src/rust/generic-subgraph-generator/src/models/file/write.rs:21]: Bit weird that this is called deleter

File: src/rust/generic-subgraph-generator/src/models/process/port_bind.rs (hunk 6 above)
Comment (d0nut-grapl) [reply src/rust/generic-subgraph-generator/src/models/process/port_bind.rs:4]: Done :D https://github.com/grapl-security/grapl/issues/307

Comment (wimax-grapl) [reply src/rust/generic-subgraph-generator/src/models/process/port_bind.rs:4]: yeah +1

File: src/rust/graph-generator-lib/src/aws.rs (hunk 10 above)
Comment (wimax-grapl) [reply src/rust/graph-generator-lib/src/aws.rs:86]: Task or //todo

File: src/rust/generic-subgraph-generator/src/serialization.rs [hunk 15]
```diff
@@ -0,0 +1,21 @@
...
+
+#[derive(Debug, Clone, Default)]
+pub struct ZstdJsonDecoder;
+
+impl<D> PayloadDecoder<D> for ZstdJsonDecoder
+where
+    for<'a> D: Deserialize<'a>,
+{
+    fn decode(&mut self, body: Vec<u8>) -> Result<D, Box<dyn std::error::Error>> {
+        let mut decompressed = Vec::new();
```
Comment (colin-grapl) [inline src/rust/generic-subgraph-generator/src/serialization.rs:13]: Minor, but I would assume that anything decompressed will be at least the size of its input, so we should just do `Vec::with_capacity(body.len())` here I think

File: src/rust/node-identifier/src/lib.rs [hunk 16]
```diff
@@ -542,7 +542,7 @@ async fn attribute_asset_ids(
     let mut output_graph = Graph::new(unid_graph.timestamp);
     output_graph.edges = unid_graph.edges;
 
-    let node_asset_ids: HashMap<String, String> = HashMap::new();
+    let _node_asset_ids: HashMap<String, String> = HashMap::new();
```
Comment (colin-grapl) [inline src/rust/node-identifier/src/lib.rs:545]: tbh with cases like this where the value is unused I'd rather it be deleted, or warn, rather than hide the warning like this.

File: src/rust/grapl-observe/src/macros.rs [hunk 17]
```diff
@@ -0,0 +1,11 @@
+#[macro_export]
+macro_rules! log_time {
+    ($msg:expr, $x:expr) => {{
+        let mut sw = stopwatch::Stopwatch::start_new();
+        #[allow(path_statements)]
+        let result = $x;
+        sw.stop();
+        info!("{} {} milliseconds", $msg, sw.elapsed_ms());
```
Comment (wimax-grapl) [inline src/rust/grapl-observe/src/macros.rs:8]: worth noting we could actually emit a histogram metric on these now, but let's save that for a next diff :)

File: src/rust/sysmon-subgraph-generator/src/serialization.rs [hunk 18]
```diff
@@ -0,0 +1,19 @@
...
+
+/// A [PayloadDecoder] used to decompress zstd encoded events sent to an [EventHandler].
+///
+/// This `struct` is typically used in conjunction with a subsequent call to [run_graph_generator].
+#[derive(Debug, Clone, Default)]
+pub struct ZstdDecoder;
+
+impl PayloadDecoder<Vec<u8>> for ZstdDecoder {
+    fn decode(&mut self, body: Vec<u8>) -> Result<Vec<u8>, Box<dyn std::error::Error>> {
+        let mut decompressed = Vec::new();
```
Comment (colin-grapl) [inline src/rust/sysmon-subgraph-generator/src/serialization.rs:12]: Let's prealloc here as well.

**Review by colin-grapl — APPROVED** [review]
Minor comments/ requests, I don't care to block on any of it really.

File: src/rust/generic-subgraph-generator/src/serialization.rs (hunk 15 above)
Comment (d0nut-grapl) [reply src/rust/generic-subgraph-generator/src/serialization.rs:13]: Good call. I'll adjust.

File: src/rust/node-identifier/src/lib.rs (hunk 16 above)
Comment (d0nut-grapl) [reply src/rust/node-identifier/src/lib.rs:545]: Yea, I'm going through node identifier currently and the majority of changes there will occur in the #306 PR

File: src/rust/sysmon-subgraph-generator/src/main.rs [hunk 19]
```diff
@@ -1,683 +1,37 @@
...
-
 #[tokio::main]
 async fn main() -> Result<(), Box<dyn std::error::Error>> {
     let env = grapl_config::init_grapl_env!();
     info!("Starting sysmon-subgraph-generator");
 
     let metrics = SysmonSubgraphGeneratorMetrics::new(&env.service_name);
 
-    if env.is_local {
+    if grapl_config::is_local() {
```
Comment (wimax-grapl) [inline src/rust/sysmon-subgraph-generator/src/main.rs:26]: can we bring this back to env.is_local?

File: src/rust/sysmon-subgraph-generator/src/models/network/inbound.rs [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+    );
+
+    // A process creates a connection
+    graph.add_edge(
+        "created_connections",
+        process.clone_node_key(),
+        outbound.clone_node_key(),
+    );
+
+    // The connection is over an IP + Port
```
Comment (wimax-grapl) [inline src/rust/sysmon-subgraph-generator/src/models/network/inbound.rs:103]: based documentation

File: src/rust/sysmon-subgraph-generator/src/models/process/create.rs [hunk 21]
```diff
@@ -0,0 +1,83 @@
...
+        .map_err(|err| failure::err_msg(err))?;
+
+    let parent = ProcessBuilder::default()
+        .asset_id(process_start.system.computer.computer.clone())
+        .state(ProcessState::Existing)
+        .process_id(process_start.event_data.parent_process_id)
+        .process_name(get_image_name(&process_start.event_data.parent_image.clone()).unwrap())
+        .process_command_line(&process_start.event_data.parent_command_line.command_line)
+        .last_seen_timestamp(timestamp)
+        //        .created_timestamp(process_start.event_data.parent_process_guid.get_creation_timestamp())
```
Comment (wimax-grapl) [inline src/rust/sysmon-subgraph-generator/src/models/process/create.rs:34]: porquoi non

File: src/rust/sysmon-subgraph-generator/src/main.rs (hunk 19 above)
Comment (d0nut-grapl) [reply src/rust/sysmon-subgraph-generator/src/main.rs:26]: yup!

File: src/rust/sysmon-subgraph-generator/src/models/process/create.rs (hunk 21 above)
Comment (d0nut-grapl) [reply src/rust/sysmon-subgraph-generator/src/models/process/create.rs:34]: This was from the original code: 

https://github.com/grapl-security/grapl/blob/staging/src/rust/sysmon-subgraph-generator/src/main.rs#L125

I figured we wanted to keep it for some reason

File: src/rust/sysmon-subgraph-generator/src/models/network/inbound.rs (hunk 20 above)
Comment (d0nut-grapl) [reply src/rust/sysmon-subgraph-generator/src/models/network/inbound.rs:103]: lol i swear that wasn't me
